metabonding/src/lib.rs

```diff
@@ -4,29 +4,36 @@ elrond_wasm::imports!();
 
 mod project;
 mod rewards;
+mod util;
 
 use core::borrow::Borrow;
-
-use elrond_wasm::api::ED25519_SIGNATURE_BYTE_LEN;
 use rewards::{RewardsCheckpoint, Week};
-
-const MAX_DATA_LEN: usize = 80; // 4 + 32 + 32, with some extra for high BigUint values
-
-pub type Signature<M> = ManagedByteArray<M, ED25519_SIGNATURE_BYTE_LEN>;
+use util::Signature;
 
 /// Source code for the pause module:
 /// https://github.com/ElrondNetwork/elrond-wasm-rs/blob/master/elrond-wasm-modules/src/pause.rs
 #[elrond_wasm::contract]
 pub trait Metabonding:
-    elrond_wasm_modules::pause::PauseModule + project::ProjectModule + rewards::RewardsModule
+    elrond_wasm_modules::pause::PauseModule
+    + project::ProjectModule
+    + rewards::RewardsModule
+    + util::UtilModule
 {
     #[init]
-    fn init(&self, signer: ManagedAddress) {
+    fn init(
+        &self,
+        signer: ManagedAddress,
+        #[var_args] opt_first_week_start_epoch: OptionalValue<u64>,
+    ) {
         self.signer().set(&signer);
         self.set_paused(true);
 
-        let current_epoch = self.blockchain().get_block_epoch();
-        self.first_week_start_epoch().set_if_empty(&current_epoch);
+        let first_week_start_epoch = match opt_first_week_start_epoch {
+            OptionalValue::Some(epoch) => epoch,
+            OptionalValue::None => self.blockchain().get_block_epoch(),
+        };
+        self.first_week_start_epoch()
+            .set_if_empty(&first_week_start_epoch);
```

Comment:
I was also referring that we should have a `require(week > current_week - PROJECT_EXPIRATION_WEEKS)` on `claimRewards` endpoint so users cannot claim rewards for one year ago for example.

Reply:
Also, claimRewards wouldn't be more suited in the rewards module? Intuitively I always look for it there before I remember it's in `lib.rs`

Reply:
The check is already done through the `is_expired` method on project, during the calculate_rewards call
```
let weekly_rewards = self.get_rewards_for_week(...)
...
if project.is_expired(current_week, true) {
                continue;
            }
...

pub fn is_expired(&self, current_week: Week, were_rewards_deposited: bool) -> bool {
        if current_week > self.end_week + PROJECT_EXPIRATION_WEEKS {
            return true;
        }
        !were_rewards_deposited && current_week >= self.start_week
    }
```

Reply:
These checks ensure that a user cannot claim rewards on an expired project. But the actual needed check is a bit stricter. A user should not be able to claim later than `PROJECT_EXPIRATION_WEEKS`, regardless of any project being expired. So if `current_week = 10`, nobody should be able to claim `week=4`. If we do this, the check for `is_expired` here is redundant.

Reply:
Done.